OvmfPkg/Library/X86QemuLoadImageLib/X86QemuLoadImageLib.c

```diff
@@ -351,11 +371,30 @@ QemuLoadKernelImage (
   Status = gBS->LoadImage (
                   FALSE,                    // BootPolicy: exact match required
                   gImageHandle,             // ParentImageHandle
-                  (EFI_DEVICE_PATH_PROTOCOL *)&mKernelDevicePath,
+                  (EFI_DEVICE_PATH_PROTOCOL *)&mShimDevicePath,
                   NULL,                     // SourceBuffer
                   0,                        // SourceSize
                   &KernelImageHandle
                   );
+  if (Status == EFI_SUCCESS) {
+    Shim = TRUE;
+    DEBUG ((DEBUG_INFO, "%a: booting via shim\n", __func__));
+  } else {
+    Shim = FALSE;
+    if (Status == EFI_SECURITY_VIOLATION) {
+      gBS->UnloadImage (KernelImageHandle);
+    }

```

Comment:
What is the logic here? If `shim` exists, was loaded successfully but did not authenticate, what is the point of carrying on?

Reply:
The point is cleaning up in case the image was loaded.  But, yes, carrying on doesn't make sense except for `Status == EFI_NOT_FOUND`.